gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoRequestUser.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.gateway.ha.router;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.airlift.log.Logger;
+import io.trino.gateway.ha.config.RequestAnalyzerConfig;
+import jakarta.servlet.http.Cookie;
+import jakarta.servlet.http.HttpServletRequest;
+import okhttp3.Call;
+import okhttp3.HttpUrl;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class TrinoRequestUser
+{
+    public static final String TRINO_USER_HEADER_NAME = "X-Trino-User";
+    public static final String TRINO_UI_TOKEN_NAME = "Trino-UI-Token";
+    public static final String TRINO_SECURE_UI_TOKEN_NAME = "__Secure-Trino-ID-Token";
+
+    private Optional<String> user = Optional.empty();
+
+    private String oauthTokenInfoUrl;
+    private final boolean isGetUserFromOAuth;
+    private static final Logger log = Logger.get(TrinoRequestUser.class);
+
+    public TrinoRequestUser(HttpServletRequest request, RequestAnalyzerConfig config)
+    {
+        isGetUserFromOAuth = config.isGetUserFromOauth();
```

Comment:
Please clarify how does this flow work, the backend trino can have multiple authentication mechanisms. so are you setting the flag for an individual request? Also gateway and trino can in theory have different IDPs.

Reply:
This is now documented under `### TrinoRequestUser` in `routing-rules.md`. Essentially it tries multiple methods, and uses whichever works. The userinfo url is configurable, it does not reuse the url from the authentication config.